CROSSTOOLS wrapped_clang: handle spaces in paths

ob opened the pull request:
When bazel calls wrapped_clang, it single-quotes all arguments. However
it passes flags with arguments quoted as a whole. That is, wrapped_clang
will be called with arguments like these:

  wrapped_clang '-isysroot /a/path/with spaces' '/a/file with spaces.m'

Before this commit, wrapped_clang was blindly splitting on space and
calling clang with invalid arguments. Now it only splits on the _first_
space, and only if the argument starts with '-'.

ob commented:
I signed in!

ob commented on tools/osx/crosstool/wrapped_clang.cc:
```diff
@@ -118,15 +118,24 @@ void RunSubProcess(const std::vector<std::string> &args) {
-// Splits txt using whitespace delimiter, pushing each substring into strs.
+// Splits txt on the first whitespace delimiter, pushing both substring into strs.
+// This is because the wrapped_clang binary is called with arguments like:
+//     wrapped_clang '-isysroot a/path/that might have/spaces' and we need to
```
Good point. I'm assuming this is called by Bazel right? I grepped for `isysroot` and it seems Bazel does use a single space:

```
tools/osx/crosstool/CROSSTOOL.tpl:        flag: "-isysroot %{sdk_dir}"
```

for example. Is it the intent to have it be general purpose and support all the different syntax? Note that this might be really hard, for instance in the `-isysroota/path` case, how do you know where the flag ends and the path starts without knowing all of clang's flags?

Also note the previous version would fail in similar or worse ways with either of these other syntaxes.

ob commented:
Thinking a bit more about this change I'm wondering if it wouldn't be better to change Bazel to *not* combine multiple arguments into one. Is there a reason it does: `wrapped_tool '-isysroot path` instead of   `wrapped_tool '-isysroot' 'path'`? That way the wrapped tools wouldn't have to parse arguments which is fragile...

ob commented:
As an update, I have found a case where this change breaks. When trying to build a sample app, I found Bazel calling the `wrapped_clang` tool with the following argument, which would need to be split into multiple arguments.

```
'-Xlinker -add_ast_path -Xlinker bazel-out/ios-x86_64-min9.0-applebin_ios-ios_x86_64-fastbuild/genfiles/sample_appBinary/_objs/sample_appBinary.swiftmodule'
```

I think what I said in [my previous comment](https://github.com/bazelbuild/bazel/pull/5147#issuecomment-387206287) is the right way forward.

ulfjack commented:
+1

ob commented:
Okay, I played a little with this but I'm stuck. The way the flags end up with spaces in them is because in `tools/osx/crosstool/CROSSTOOL.tpl` they are specified as:

```
       flag_group {
         flag: "-isysroot %{sdk_dir}"
       }
```

I've verified that if I remove that space and make the flag two entries, like so:

```
       flag_group {
         flag: "-isysroot"
         flag: "%{sdk_dir}"
       }
```

The problem is solved. However that file says at the beginning:

```
# This file was auto-generated by a script maintained internally by
# Google.
```

Which makes me thing that changing it directly is a bad idea. Also, there are other instances of spaces in the flags, but there is some kind of iteration going on, e.g. 

```
      flag_group {
        flag: "-weak_framework %{weak_framework_names}"
        iterate_over: "weak_framework_names"
      }
```

I am not familiar with the language specification for this file so I don't know how that could be turned into multiple `flag` entries that are paired with their arguments.

ob commented:
Those last commits kind of work for me but I'm not sure whether that `iterate_over` iterates over the `flag_group` or single flags... or whether this needs to be changed in some internal Google script.

ulfjack commented:
tools/osx/crosstool/CROSSTOOL.tpl changes require a change in an internal Google script, correct. Is the patch in this CL the resulting file we need to get?

ulfjack commented:
I have a change to update the script that I'm about to send out. It should get referenced here once it's merged.

ob commented:
Yeah, the changes to CROSSTOOL.tpl are what I'm testing with. Modulo [my previous comments](https://github.com/bazelbuild/bazel/pull/5147#issuecomment-392942413) about my lack of knowledge of tpl format and whether it expands flag_groups as I _think_ it expands them, then yeah.

Note that I also removed the parsing from the clang wrapper. 

Thank you for taking a look at this.

ulfjack commented:
My change was merged. Please rebase and let me know whether you can get it to work now.

ob commented:
@ulfjack I tested this with my workspace and the test case that I had added before. Note that I removed splitting on spaces that `wrapped_clang` used to do.

ulfjack commented:
Generally looks good. I triggered the CI.

ob commented:
I should probably fix the comments before you merge and reference the commit you did that fixes the caller side...

ulfjack commented:
@ob, you can't trigger CI yourself. Security told us we're required to do a code review first.

ob commented:
@ulfjack that makes sense... in a way triggering a CI on unreviewed code changes can be seen as a remote execution attack ;-)

ulfjack commented:
Triggered.

ob commented:
Hmm... the test failure seems unrelated to my changes... FWIW, ```//src/test/java/com/google/devtools/build/lib:remote-tests``` passes on my machine.

ulfjack commented:
Yeah, the test is flaky - I just disabled it earlier today. I'm going to merge this PR now.
